### PR TITLE
[Inductor] Generalize `is_cuda` to specific device_type to make cpp_wrapper mode be extensible

### PR DESCRIPTION
**Pull request by etaf**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #135320
* #135318
* #135313
* #135312
* #134761
* __->__ #134693

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang

### CODE REVIEW

File: torch/_inductor/codecache.py [hunk 1]
```diff
@@ -1547,7 +1546,9 @@ def compile(
         use_absolute_path = False
         if config.is_fbcode():
             ld_command = build_paths.ld()
-            if not cuda and graph.aot_mode:  # Meta internal AOTInductor CPU
+            if (
+                not (device_type == "cuda") and graph.aot_mode
```
Comment (ezyang) [inline torch/_inductor/codecache.py:1550]: Based on the comment here, it seems more appropriate to test `device_type == "cpu"` here

File: torch/_inductor/codecache.py [hunk 2]
```diff
@@ -2696,7 +2700,7 @@ def generate_halide_async(
             cls._codegen_glue(meta, headerfile),
             extra_flags=(libfile, cls.build_standalone_runtime()),
             submit_fn=jobs.append if need_compile else None,
-            cuda=meta.is_cuda(),
+            device_type="cuda" if meta.is_cuda() else "cpu",
```
Comment (ezyang) [inline torch/_inductor/codecache.py:2703]: You don't need to generalize the meta struct?

File: torch/_inductor/codegen/cpp_wrapper_cpu.py [hunk 3]
```diff
@@ -1717,7 +1721,11 @@ def create_dtypeview_call(reinterpret_call):
                 )
                 call_strs = [f"AtenTensorHandle {tmp_AtenTensorHandle};"]
                 dtype_name = str(dtype).split(".")[-1]
-                device_name = "cuda" if data.layout.device.type == "cuda" else "cpu"
+                device_name = (
+                    data.layout.device.type
+                    if is_gpu(data.layout.device.type)
+                    else "cpu"
```
Comment (ezyang) [inline torch/_inductor/codegen/cpp_wrapper_cpu.py:1727]: why not just unconditionally data.layout.device.type here?

**Comment by ezyang** [comment]
This seems basically fine. Draft?

File: torch/_inductor/codecache.py (hunk 1 above)
Comment (etaf) [reply torch/_inductor/codecache.py:1550]: Sure, thanks.

File: torch/_inductor/codecache.py (hunk 2 above)
Comment (etaf) [reply torch/_inductor/codecache.py:2703]: Here, the meta is a `HalidMeta`.  I was originally thinking of keeping the changes minimal for now and waiting to generalize this when Halide backend is supported on Intel GPU. What do you think?

File: torch/_inductor/codegen/cpp_wrapper_cpu.py (hunk 3 above)
Comment (etaf) [reply torch/_inductor/codegen/cpp_wrapper_cpu.py:1727]: > why not just unconditionally data.layout.device.type here?

Sure, the origin `device_name = "cuda" if data.layout.device.type == "cuda" else "cpu"` also confused me.

File: torch/_inductor/codecache.py (hunk 2 above)
Comment (ezyang) [reply torch/_inductor/codecache.py:2703]: ok

**Review by ezyang — APPROVED** [review]

**Review by EikanWang — APPROVED** [review]

**Review by jansel — APPROVED** [review]

**Pull request closed without merging**